### PR TITLE
Exercise

### PR DESCRIPTION
**Pull request by allhaker**
Check the readme for more info. It's also in the PR. I did it so that you can distinguish easier between the template and my work.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (allhaker) [inline .eslintrc.json:1]: I added linter config with prettier.
Config is may be not the best but I didn't want to spend much time here.
And I refuse to work without Linter =) In 2023 this should be illegal.

File: .nvmrc [hunk 2]
```diff
@@ -0,0 +1 @@
+18.12
```
Comment (allhaker) [inline .nvmrc:1]: I always use nvm.

File: package.json [hunk 3]
```diff
@@ -28,6 +30,12 @@
       "react-app/jest"
     ]
   },
+  "jest": {
```
Comment (allhaker) [inline package.json:33]: Jest was complaining without this one. Modules were not transpiling.

File: src/App.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+const apiConfig = { BATCH_DUMP_TIMEOUT_MS: 1, BATCH_URL: '/file-batch-api' };
+jest.mock('./api-client/consts', () => apiConfig);
+
+describe('App component', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test('fetches and displays files correctly', async () => {
```
Comment (allhaker) [inline src/App.spec.tsx:15]: This could use a test where we could see that everything works with multiple batch dumps.
But I was lazy. Already spent quite a lot of time.

Comment (allhaker) [reply src/App.spec.tsx:15]: If it were production code I'd have added such a test.

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,78 @@
-import React from 'react';
+import React, { useCallback, useState } from 'react';
 import logo from './logo.svg';
 import './App.css';
+import { BATCH_DUMP_TIMEOUT_MS, FilesApi } from './api-client';
+
+const fileIds = ['fileid1', 'fileid2', 'fileid3'];
+
+const App = () => {
+  let combinedFiles: string[] | undefined;
```
Comment (allhaker) [inline src/App.tsx]: I am not proud of this. Because requestFile gets promise response at the same time you can't easily combine files payload. I decided to go with this. So I have this variable outside of the closure and use it as an intermediary to combine files. On every render it's reset.

In a real life it's possible that we would have had a context and data combination would have been handled there. I thought about context but decided that for such a small app it's an overkill.

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,78 @@
...
+    },
+    [files]
+  );
 
-function App() {
   return (
     <div className="App">
       <header className="App-header">
         <img src={logo} className="App-logo" alt="logo" />
+        <h2>Let&apos;s make some batch requests</h2>
```
Comment (allhaker) [inline src/App.tsx:41]: This could have probably been split into a couple of components. But they would have been quite small and wouldn't have had much logic. So I just did it all in one file.

File: src/api-client/files-api.spec.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+jest.mock('./api-client', () => ({
+  ApiClient: {
+    get: jest.fn()
+  }
+}));
+
+const defaultFiles = ['fileid1', 'fileid2', 'fileid3'];
+
+describe('Files API', () => {
```
Comment (allhaker) [inline src/api-client/files-api.spec.ts:15]: This test suite does have quite a few tests. I think I tested everything that came to my head.

File: src/App.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import React from 'react';
+import { render, fireEvent, waitFor } from '@testing-library/react';
+import App from './App';
+
+const WAIT_TIMEOUT_MS = 3000;
+
+const apiConfig = { BATCH_DUMP_TIMEOUT_MS: 1, BATCH_URL: '/file-batch-api' };
+jest.mock('./api-client/consts', () => apiConfig);
+
+describe('App component', () => {
```
Comment (allhaker) [inline src/App.spec.tsx:10]: I should have mocked the endpoint like in the other test.
We want unit tests to be fast and reliable.

So, for a production use I would have either moved this test to `intergration-tests` directory or mocked the api responses.

File: src/api-client/files-api.spec.ts (hunk 7 above)
Comment (allhaker) [reply src/api-client/files-api.spec.ts:15]: I also tested quite a few of corner case and error scenarios.

File: src/api-client/files-api.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+type RequestResolveFn = (value: AxiosFileResponse) => void;
+
+interface BatchRequest {
+  requestPromise: Promise<AxiosFileResponse>;
+  requestResolveFn: RequestResolveFn;
+  requestRejectFn: RequestResolveFn;
+  ids: string[];
+}
+
+const getClientWithBatching = () => {
```
Comment (allhaker) [inline src/api-client/files-api.ts:21]: I decided against using an interceptor what was discussed in the exercise text.
1. That approach is very non React. I think interceptors are more of a backend thing and that would have been very hidden from devs and IMHO unintended mistakes are likely
2. Completely global with interceptors. What if you want to send requests without batching? Then you need to pass some parameters via request... Dirty in my opinion.
3. I would have had to use `adapter` and use the guts of the `axios`. I don't think that is the safest and most reliable way to achieve this result.

So, all things combined I just did it with promises. Intercepting requests, returning intermediate promises, combining payloads, sending 1 actual request and then resolving correct promises with correct data.

File: src/api-client/files-api.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+  const getBatchRequest = (ids: string[]) => {
+    let requestResolveFn: RequestResolveFn | undefined;
+    let requestRejectFn: RequestResolveFn | undefined;
+    const requestPromise = new Promise<AxiosFileResponse>((resolve, reject) => {
+      requestResolveFn = resolve;
+      requestRejectFn = reject;
+    });
+
+    if (!requestResolveFn || !requestRejectFn) throw new Error('Application Internal error');
```
Comment (allhaker) [inline src/api-client/files-api.ts:32]: This is just to shut up TS. This error will never actually happen.
